cpp/src/groupby/sort/group_rank_scan.cu

```diff
@@ -93,53 +123,212 @@ std::unique_ptr<column> rank_generator(column_view const& order_by,
+    },
+    [] __device__(auto val1, auto val2) {
+      return val1 == 0 or val2 == 0 ? std::max(val1, val2) : std::min(val1, val2);
+    },
+    has_nested_nulls(table_view{{grouped_values}}),
+    stream,
+    mr);
+}
+
+std::unique_ptr<column> first_rank_scan(column_view const& grouped_values,
+                                        column_view const&,
+                                        device_span<size_type const> group_labels,
+                                        device_span<size_type const> group_offsets,
+                                        rmm::cuda_stream_view stream,
+                                        rmm::mr::device_memory_resource* mr)
+{
+  auto ranks = make_fixed_width_column(
+    data_type{type_to_id<size_type>()}, group_labels.size(), mask_state::UNALLOCATED, stream, mr);
+  auto mutable_ranks = ranks->mutable_view();
+  thrust::tabulate(
+    rmm::exec_policy(stream),
+    mutable_ranks.begin<size_type>(),
+    mutable_ranks.end<size_type>(),
+    [labels = group_labels.data(), offsets = group_offsets.data()] __device__(size_type row_index) {
+      auto group_start = offsets[labels[row_index]];
+      return row_index - group_start + 1;
+    });
+  return ranks;
+}
+
+std::unique_ptr<column> average_rank_scan(column_view const& grouped_values,
+                                          column_view const& value_order,
+                                          device_span<size_type const> group_labels,
+                                          device_span<size_type const> group_offsets,
+                                          rmm::cuda_stream_view stream,
+                                          rmm::mr::device_memory_resource* mr)
+{
+  auto max_rank = max_rank_scan(grouped_values,
+                                value_order,
+                                group_labels,
+                                group_offsets,
+                                stream,
+                                rmm::mr::get_current_device_resource());
+  auto min_rank = min_rank_scan(grouped_values,
+                                value_order,
+                                group_labels,
+                                group_offsets,
+                                stream,
+                                rmm::mr::get_current_device_resource());
+  auto ranks    = make_fixed_width_column(
+    data_type{type_to_id<double>()}, group_labels.size(), mask_state::UNALLOCATED, stream, mr);
+  auto mutable_ranks = ranks->mutable_view();
+  thrust::transform(rmm::exec_policy(stream),
+                    max_rank->view().begin<size_type>(),
+                    max_rank->view().end<size_type>(),
+                    min_rank->view().begin<size_type>(),
+                    mutable_ranks.begin<double>(),
+                    [] __device__(auto max_rank, auto min_rank) -> double {
+                      return min_rank + (max_rank - min_rank) / 2.0;
+                    });
```

Comment:
I guess there's no way to avoid this multi-pass approach of computing the min and max scans first and then doing a third transform here because we need to establish the bounds of each group before we can compute an average and there's no easy way to accumulate those in a single pass? I was hoping to see something like this operator embedded deeper in the call stack so that it could just use a single kernel, but I'm assuming you considered that and there just isn't an easy solution here.

Reply:
There is another way to do it in single pass. I used it in `cudf::rank`, it involves dense_rank computation, `reduce_by_key`, and `scatter`, which is 2 same calculations with extra work.

Even if I could do this, it would involve code duplication anyway.